chore: #1677 Rename LetterSoundCorrespondence to LetterSound

SnehaHS65 opened the pull request:
This pull request addresses issue #1677 by renaming the class LetterSoundCorrespondence to LetterSound and updating all references throughout the codebase.

nya-elimu commented on src/main/java/ai/elimu/model/content/Word.java:
```diff
@@ -48,12 +48,12 @@ public void setText(String text) {
     
-    public List<LetterSoundCorrespondence> getLetterSoundCorrespondences() {
-        return letterSoundCorrespondences;
+    public List<LetterSound> getLetterSoundCorrespondences() {
```
```suggestion
    public List<LetterSound> getLetterSounds() {
```

venkatesh2k3 commented on src/main/java/ai/elimu/dao/LetterSoundContributionEventDao.java:
```diff
@@ -10,7 +10,7 @@ public interface LetterSoundContributionEventDao extends GenericDao<LetterSoundC
     List<LetterSoundCorrespondenceContributionEvent> readAllOrderedByTimeDesc() throws DataAccessException;
 
-    List<LetterSoundCorrespondenceContributionEvent> readAll(LetterSoundCorrespondence letterSound) throws DataAccessException;
+    List<LetterSoundCorrespondenceContributionEvent> readAll(LetterSound letterSound) throws DataAccessException;
```
`LetterSoundCorrespondenceContributionEvent` to be changed to `LetterSoundContributionEvent`

venkatesh2k3 commented on src/main/java/ai/elimu/dao/jpa/LetterSoundContributionEventDaoJpa.java:
`LetterSoundCorrespondenceContributionEvent` in SQL query to be changed

venkatesh2k3 commented on src/main/java/ai/elimu/dao/jpa/LetterSoundDaoJpa.java:
`String letterSoundCorrespondenceLetters` look at the variables declared.

venkatesh2k3 commented on src/main/java/ai/elimu/dao/jpa/LetterSoundPeerReviewEventDaoJpa.java:
Even change the SQL queries.

venkatesh2k3 commented on src/main/java/ai/elimu/logic/converters/StringToLetterSoundCorrespondenceConverter.java:
`letterSoundCorrespondenceId` change this to `letterSoundId`

venkatesh2k3 commented on src/main/java/ai/elimu/model/content/Word.java:
```diff
@@ -48,12 +48,12 @@ public void setText(String text) {
 
-    public void setLetterSoundCorrespondences(List<LetterSoundCorrespondence> letterSoundCorrespondences) {
-        this.letterSoundCorrespondences = letterSoundCorrespondences;
+    public void setLetterSoundCorrespondences(List<LetterSound> letterSounds) {
```
Also change setters and getters

venkatesh2k3 commented on src/main/java/ai/elimu/model/contributor/LetterSoundCorrespondenceContributionEvent.java:
```diff
@@ -10,13 +10,13 @@ public class LetterSoundCorrespondenceContributionEvent extends ContributionEven
 
-    public LetterSoundCorrespondence getLetterSoundCorrespondence() {
-        return letterSoundCorrespondence;
+    public LetterSound getLetterSoundCorrespondence() {
```
Make sure to change getters and setters.

venkatesh2k3 commented on src/main/java/ai/elimu/rest/v2/JpaToGsonConverter.java:
```diff
@@ -129,7 +123,7 @@ public static WordGson getWordGson(Word word) {
             wordGson.setText(word.getText());
             List<LetterSoundGson> letterSounds = new ArrayList<>();
-            for (LetterSoundCorrespondence letterSound : word.getLetterSoundCorrespondences()) {
+            for (LetterSound letterSound : word.getLetterSoundCorrespondences()) {
```
getter method to be changed

venkatesh2k3 commented on src/main/java/ai/elimu/tasks/LetterSoundCorrespondenceUsageCountScheduler.java:
```diff
@@ -40,28 +40,28 @@ public synchronized void execute() {
-                        letterSoundCorrespondenceFrequencyMap.getOrDefault(letterSoundCorrespondence.getId(), 0) + word.getUsageCount());
+            for (LetterSound letterSound : word.getLetterSoundCorrespondences()) {
+                letterSoundCorrespondenceFrequencyMap.put(letterSound.getId(),
+                        letterSoundCorrespondenceFrequencyMap.getOrDefault(letterSound.getId(), 0) + word.getUsageCount());
```
`letterSoundCorrespondenceFrequencyMap` change it

venkatesh2k3 commented on src/main/java/ai/elimu/tasks/SoundUsageCountScheduler.java:
```diff
@@ -42,14 +42,14 @@ public synchronized void execute() {
-            for (LetterSoundCorrespondence letterSoundCorrespondence : word.getLetterSoundCorrespondences()) {
-                for (Sound sound : letterSoundCorrespondence.getSounds()) {
-                    soundFrequencyMap.put(sound.getId(), soundFrequencyMap.getOrDefault(sound.getId(), 0) + letterSoundCorrespondence.getUsageCount());
+            for (LetterSound letterSound : word.getLetterSoundCorrespondences()) {
```
`getLetterSoundCorrespondences()`

venkatesh2k3 commented on src/main/java/ai/elimu/util/csv/CsvContentExtractionHelper.java:
```diff
@@ -56,10 +53,10 @@ public class CsvContentExtractionHelper {
      */
-    public static List<LetterSoundCorrespondence> getLetterSoundCorrespondencesFromCsvBackup(File csvFile, LetterDao letterDao, SoundDao soundDao, LetterSoundDao letterSoundDao) {
+    public static List<LetterSound> getLetterSoundCorrespondencesFromCsvBackup(File csvFile, LetterDao letterDao, SoundDao soundDao, LetterSoundDao letterSoundDao) {
         logger.info("getLetterSoundCorrespondencesFromCsvBackup");
```
`getLetterSoundCorrespondencesFromCsvBackup`

venkatesh2k3 reviewed (changes requested):
Getter and Setter methods to be changed and variables declared, The method return type to be changed in each file

SnehaHS65 commented:
@venkatesh2k3 , I have changed in all the java files as you mentioned, should I change in webapp folder(in html and css files) as well? Let me know and I can commit

jo-elimu reviewed (changes requested):
@SnehaHS65 Thank you for opening this pull request 🙂

I will have to ask you to revert the second commit in your pull request: https://github.com/elimu-ai/webapp/pull/1682/commits/35093efec841f6c7b2b8233c732bc090ce548afc

This is because:

- The scope for the changes should only be for `LetterSoundCorrespondence.java`, and not for additional entities. (The number of code changes is already high in the initial commit, so we shouldn't add additional complexity.)
- The changes are touching the database layer, which means that we will also have to create a database migration script for your changes. See [DB_MIGRATION.md](https://github.com/elimu-ai/webapp/blob/main/DB_MIGRATION.md).
- A database schema export file is missing. You can generate it by running `mvn jetty:run` to launch the webapp locally. You will then see code changes in the `jpa-schema-export.sql` file, and those changes should also be included in your PR.
    ![](https://github.com/elimu-ai/webapp/assets/1451036/f589d751-9250-41d1-ba77-3239072bc193)

> [!NOTE]
> The reason why we need to create a DB migration script is because we already have a MySQL production database. So anytime we make changes to the database structure, we have to write SQL scripts for updating the database. See [DB_MIGRATION.md](https://github.com/elimu-ai/webapp/blob/main/DB_MIGRATION.md).

> [!TIP]
> When you create the database migration script, you can look at the changes in `jpa-schema-export.sql` to know which tables/columns to modify.

As this PR is already quite large, you might want to split it up into smaller parts to make sure that it doesn't break the database structure. For example one PR for only renaming the `LetterSoundCorrespondence` database table, and another PR for renaming  the `LetterSoundCorrespondence_id` references another database table, etc.

As you can see in the `jpa-schema-export.sql` diff, it gets quite complicated with all the changes to the database column names, primary keys, and foreign key constraints. And that is why I recommend we reduce the size of your PR.

SnehaHS65 commented:
@jo-elimu , Can you review the latest commit now, I refactored `LetterSoundCorrespondence.java` to `LetterSound.java`, ran this command `mvn jetty:run`, included `jpa-schema-export.sql` in the PR and also added DB script for the same.

SnehaHS65 commented:
Hi @jo-elimu , @nya-elimu , @venkatesh2k3 , can you please look at the latest changes that I did to rename LetterSoundCorrespondence to LetterSound. This is the first PR to just rename the LetterSoundCorrespondence database table as suggested by @jo-elimu. If you approve this, I can raise one more PR for renaming the LetterSoundCorrespondence_id references.

jo-elimu commented:
> Hi @jo-elimu , @nya-elimu , @venkatesh2k3 , can you please look at the latest changes that I did to rename LetterSoundCorrespondence to LetterSound. This is the first PR to just rename the LetterSoundCorrespondence database table as suggested by @jo-elimu. If you approve this, I can raise one more PR for renaming the LetterSoundCorrespondence_id references.

@SnehaHS65 If you look at your changes in [`jpa-schema-export.sql`](https://github.com/elimu-ai/webapp/pull/1682/files#diff-94be1c5456b98ad8b18c1bf1dfb56ed60c79d2c490d5c12d93a1d99dc77fc48d), you'll see that it contains around 16 changes, and these 16 changes would also have to be included in the DB migration script.

This is why I suggested above that we split #1677 into smaller pull requests to make each database migration less complicated.

I've created an example here: https://github.com/elimu-ai/webapp/pull/1713. This pull request only makes one small change (renaming from `Word#letterSoundCorrespondences` to `Word#letterSoundCorrespondences`). After merging that PR, we can perform a release and check if the migration script runs without errors. If so, we can move on to another small part of the issue.

SnehaHS65 commented:
> > Hi @jo-elimu , @nya-elimu , @venkatesh2k3 , can you please look at the latest changes that I did to rename LetterSoundCorrespondence to LetterSound. This is the first PR to just rename the LetterSoundCorrespondence database table as suggested by @jo-elimu. If you approve this, I can raise one more PR for renaming the LetterSoundCorrespondence_id references.
> 
> @SnehaHS65 If you look at your changes in [`jpa-schema-export.sql`](https://github.com/elimu-ai/webapp/pull/1682/files#diff-94be1c5456b98ad8b18c1bf1dfb56ed60c79d2c490d5c12d93a1d99dc77fc48d), you'll see that it contains around 16 changes, and these 16 changes would also have to be included in the DB migration script.
> 
> This is why I suggested above that we split #1677 into smaller pull requests to make each database migration less complicated.
> 
> I've created an example here: #1713. This pull request only makes one small change (renaming from `Word#letterSoundCorrespondences` to `Word#letterSounds`). After merging that PR, we can perform a release and verify that the migration script runs without errors. If so, we can move on to another small part of the issue.

@jo-elimu ,So there are ContributionEvent and PeerReviewEvent which uses this letterSoundCorrespondence, I will now only edit for ContributionEvent first and commit?

jo-elimu commented:
> @jo-elimu ,So there are ContributionEvent and PeerReviewEvent which uses this letterSoundCorrespondence, I will now only edit for ContributionEvent first and commit?

@SnehaHS65 Yes, that's right :-)

I suggest you close this pull request, and open a new pull request for renaming the `LetterSoundCorrespondenceContributionEvent` class. And remember to include the DB migration script as well (see similar example in [`2004010.sql`](https://github.com/elimu-ai/webapp/blob/main/src/main/resources/db/migration/2004010.sql)).
